### PR TITLE
Fix async tests

### PR DESCRIPTION
**Pull request by RobLoach**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -134,9 +142,13 @@ function addTests(transform, testDirectory, test) {
...
 
     if (transform.renderAsync) {
-      test(transform.name + '.renderAsync()', function () {
+      test(transform.name + '.renderAsync()', function (done) {
         return transform.renderAsync(input, options, locals).then(function (output) {
           checkOutput(output);
+          done();
+        }, function (err) {
+          throw new Error(err);
+          done(new Error(err));
```
Comment (RobLoach) [inline index.js]: @ForbesLindesay `done(err)` was not reporting failiures. Try this in `test/index.js`...

``` js
testJStransformer({
  name: 'renderAsync: Expect Failure',
  outputFormat: 'txt',
  inputFormats: ['txt'],
  renderAsync: function (str, options, locals) {
    return new Promise(function(fulfill, reject) {
      reject('FAILURE');
    });
  }
}, resolve(__dirname + '/../example/simple'));
```

Know what's wrong with it?

**Comment by ForbesLindesay** [comment]
Testit has built in promise support.  This shouldn't be needed.

**Comment by RobLoach** [comment]
> @ForbesLindesay Testit has built in promise support. This shouldn't be needed.

Not sure if it's a problem with testit, or our usage of it, but I've committed [the broken test that `rejects()`](https://github.com/jstransformers/test-jstransformer/blob/da0080e58a79c65d1bf7f6fbf2e498bdb0aa4464/test/index.js#L80L93), but yet [seems to pass](https://travis-ci.org/jstransformers/test-jstransformer/jobs/66318367#L153-L157).

**Comment by RobLoach** [comment]
> Testit has built in promise support. This shouldn't be needed.

We still need to check whether or not the output is correct when the asyncronous call is complete, however.

I've updated it to catch thrown Errors, and not return the Promise since we're using `done` instead. Should be good now!

**Comment by TimothyGu** [comment]
Just as @ForbesLindesay said, testit supports Promises just fine. To prove that, I've whipped up a [new branch](https://github.com/jstransformers/test-jstransformer/compare/async-tmp) freshly cut from master with your should-be-failing test, and it [fails correctly](https://travis-ci.org/jstransformers/test-jstransformer/jobs/70176478#L152).

I suspect that the cause of the test passing in your case is because of the commit that came before the commit that added the failure test, c9138dd.

**Comment by RobLoach** [comment]
I believe what they were missing is the `.catch(done)`. I was working on the SASS one, and it wasn't building, but was reporting passes. Having this makes it so that if the transformer throws an Error, it catches it and reports it back to the Promise correctly.

**Comment by TimothyGu** [comment]
@RobLoach I don't think `.catch` works the way you think it does. An error in the Promise propagates down the promise chain, so it doesn't stop as soon as you `return` and needs an extra `.catch(done)`. The fact that you are handling the error in multiple places is against the principles of Promise/A+.

But anyway, is the SASS test you are having trouble with publicly available? If so I'll take a look at that specific case instead.

**Comment by RobLoach** [comment]
@TimothyGu https://github.com/jstransformers/jstransformer-nunjucks/pull/3 .... Not sure if returning a Promise for a Promise is a good approach, but it demonstrates a break when using test-jstransformer 1.0.1. Could probably be misuse of it though.

Considering just removing .compileAsync entirely for that one.

**Comment by TimothyGu** [comment]
You're right that the returning Promise => (locals) => Promise doesn't work with jstransformer. It is not designed to be used this way, and thus test-jstransformer behaves oddly. This is a jstransformer deficiency.

**Comment by RobLoach** [comment]
@TimothyGu What does compileAsync expect? Particularily when desired to function on renderAsync?

**Comment by TimothyGu** [comment]
compileAsync doesn't fallback on renderAsync.

Some examples for each:

``` js
var tr = require('jstransformer')(require('jstransformer-jade'))

tr.renderAsync('p= greeting', {}, {greeting: 'hey'}).done(function (output) {
  assert.equal(output.body, '<p>hey</p>')
})

tr.compileAsync('p= greeting', {}).done(function (output) {
  assert.equal(output.fn({greeting: 'hey'}), '<p>hey</p>')
})

// dust/nunjacks-style: DOESN'T WORK

tr.compileAsync('p= greeting', {}).then(function (output) {
  output.fn({greeting: 'hey'}).then(function (body) {
    assert.equal(body, '<p>hey</p>')
  })
})
```

**Comment by ForbesLindesay** [comment]
Since there doesn't seem to be a bug with test-jstransformer I'm going to close this pull reuqest.

I do think we should consider either adding:
- `compileToAsync`
- `compileFileToAsync`
- `compileAsyncToAsync`
- `compileFileAsyncToAsync`

which would support pre-compiled async render and would fall back on their respective non-`ToAsync` versions.  It was an oversight in the initial API design that this case was omitted.

**Pull request closed without merging**